Update discovery journal on publish

tuchfarber opened the pull request.

bill-filler commented on journals/apps/journals/handlers.py:
```diff
@@ -1,15 +1,26 @@
-page_unpublished.connect(unpub_receiver, sender=JournalPage)
+def about_page_pub_receiver(sender, **kwargs):
+    journal_about_age = kwargs['instance']
+    journal_about_age.update_related_objects()
```
shouldn't these be "journal_about_page" not "journal_about_age"?
assuming typo

tuchfarber replied:
Typo, will fix

bill-filler commented on journals/apps/journals/handlers.py:
```diff
@@ -1,15 +1,26 @@
+
+def about_page_unpub_receiver(sender, **kwargs):
+    journal_about_age = kwargs['instance']
+    journal_about_age.update_related_objects(deactivate=True)
```
same comment about name as above

tuchfarber replied:
Typo, will fix

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -126,7 +126,8 @@ def get_discovery_data(self):
             'short_description': self.journal_about_page.short_description,
-            'full_description': self.journal_about_page.long_description
+            'full_description': self.journal_about_page.long_description,
+            'status': 'active' if self.journal_about_page.live else 'inactive',
```
just a question:
might this field be better as "active" with value of True|False?
Then we wouldn't have to compare string values when checking the status but instead just testing boolean
I'm fine either way, just a suggestion

tuchfarber replied:
I was basing it off of the way Courses and Programs has status. It also allows for us to add more options down the road if need be.

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +315,17 @@ def generate_basket_url(self, sku):
+    def update_related_objects(self, deactivate=False):
+        data = {
+            "status": "active" if not deactivate else "inactive",
+            "card_image_url": self.site.root_url + self.card_image.file.url,
```
how, if any, will this change when we use S3 backend for assests?

tuchfarber replied:
Hmmm, good point. We definitely won't want to build the URL this way. I'll see if I can find a way to get a FQDN URL.

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +315,17 @@ def generate_basket_url(self, sku):
+    def update_related_objects(self, deactivate=False):
+        data = {
+            "status": "active" if not deactivate else "inactive",
+            "card_image_url": self.site.root_url + self.card_image.file.url,
```
just got an error when running the publish_journals mgmt command:
```
 File "/edx/app/journals/journals/journals/apps/journals/models.py", line 327, in update_related_objects
    "card_image_url": self.site.root_url + self.card_image.file.url,
AttributeError: 'NoneType' object has no attribute 'file'
```
We could check for None but might be better if we set a default card image on the about page as a default value. That way something will show up in the card on the index page and in LMS card image for the Journal

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +315,17 @@ def generate_basket_url(self, sku):
+            "short_description": self.short_description,
+        }
+        discovery_journal_api_client = self.site.siteconfiguration.discovery_journal_api_client
+        discovery_journal_api_client.journals(self.journal.uuid).patch(data)
```
So on initial run of publish_journal mgmt command this ends up being called because the about page is being published https://github.com/edx/journals/blob/master/journals/apps/journals/management/commands/publish_journals.py#L115 but it's not necessary to modify it here as we've just set it during creation and nothing has changed. 
Wonder if there is a way to only call the patch if the data is different, or somehow ignore the publish if it was done from the mgmt command?

bill-filler replied:
Actually, this is a problem. During the mgmt command, the wagtail pages are created first before the journal data is pushed to discovery. Your .patch() is failing because the journal with specified uuid does not yet exist:
```
root@00334e2d7247:/edx/app/journals/journals# ./manage.py publish_journals --create "Boo-123" --org "edx" --price "200.00" --access_length "180"
2018-04-12 03:17:47,974 INFO 72 [elasticsearch] base.py:83 - PUT http://journals.elasticsearch:9200/wagtail__wagtailcore_page/wagtailcore_page_journals_journalaboutpage/wagtailcore_page%3A53 [status:201 request:0.018s]
2018-04-12 03:17:47,976 INFO 72 [wagtail.core] models.py:503 - Page created: "Boo-123 About Page" id=53 content_type=journals.JournalAboutPage path=/mitxpro-index-page/boo-123-about-page/
2018-04-12 03:17:48,030 INFO 72 [elasticsearch] base.py:83 - PUT http://journals.elasticsearch:9200/wagtail__wagtailcore_page/wagtailcore_page_journals_journalaboutpage/wagtailcore_page%3A53 [status:200 request:0.016s]
2018-04-12 03:17:48,039 INFO 72 [wagtail.core] models.py:685 - Page edited: "Boo-123 About Page" id=53 revision_id=52
2018-04-12 03:17:48,078 INFO 72 [elasticsearch] base.py:83 - PUT http://journals.elasticsearch:9200/wagtail__wagtailcore_page/wagtailcore_page_journals_journalaboutpage/wagtailcore_page%3A53 [status:200 request:0.016s]
Traceback (most recent call last):
  File "./manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/edx/app/journals/journals/journals/apps/journals/management/commands/publish_journals.py", line 160, in handle
    journal = self._handle_create(options)
  File "/edx/app/journals/journals/journals/apps/journals/management/commands/publish_journals.py", line 45, in _handle_create
    journal_about_page = self._update_wagtail_pages(journal, create=True, publish=publish)
  File "/edx/app/journals/journals/journals/apps/journals/management/commands/publish_journals.py", line 115, in _update_wagtail_pages
    journal_about_page.save_revision().publish()
  File "/usr/local/lib/python3.6/site-packages/wagtail/wagtailcore/models.py", line 1524, in publish
    page_published.send(sender=page.specific_class, instance=page.specific, revision=self)
  File "/usr/local/lib/python3.6/site-packages/django/dispatch/dispatcher.py", line 193, in send
    for receiver in self._live_receivers(sender)
  File "/usr/local/lib/python3.6/site-packages/django/dispatch/dispatcher.py", line 193, in <listcomp>
    for receiver in self._live_receivers(sender)
  File "/edx/app/journals/journals/journals/apps/journals/handlers.py", line 16, in about_page_pub_receiver
    journal_about_age.update_related_objects()
  File "/edx/app/journals/journals/journals/apps/journals/models.py", line 333, in update_related_objects
    discovery_journal_api_client.journals(self.journal.uuid).patch(data)
  File "/usr/local/lib/python3.6/site-packages/slumber/__init__.py", line 171, in patch
    resp = self._request("PATCH", data=data, files=files, params=kwargs)
  File "/usr/local/lib/python3.6/site-packages/slumber/__init__.py", line 101, in _request
    raise exception_class("Client Error %s: %s" % (resp.status_code, url), response=resp, content=resp.content)
slumber.exceptions.HttpNotFoundError: Client Error 404: http://edx.devstack.discovery:18381/journal/api/v1/journals/6303f0c8-00c4-424e-9a27-71ff52454688/
```

tuchfarber commented:
All issues should be taken care of now. I added a property to get absolute URL of card image to solve the S3 vs local issue. Also fixed API call error handling.

Will squash post review.

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +317,35 @@ def generate_basket_url(self, sku):
+
+        try:
+            discovery_journal_api_client.journals(self.journal.uuid).patch(data)
+        except HttpNotFoundError as err:
```
this needs to be imported

tuchfarber replied:
Will fix

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +317,35 @@ def generate_basket_url(self, sku):
+            logging.error("Error updating discovery after JournalAboutPage publish: {}".format(err.content))
+
+    @property
+    def card_image_absolute_url():
```
should be
```
def card_image_absolute_url(self):
```

tuchfarber replied:
Will fix

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -314,6 +317,35 @@ def generate_basket_url(self, sku):
+        data = {
+            "status": "active" if not deactivate else "inactive",
+            "card_image_url": self.card_image_absolute_url,
+            "title": self.title,
```
As the title is used in ecommerce (on the basket page), we need to update ecommerce as well when the title changes

tuchfarber replied:
I'll write the code for it, but the current patch endpoint requires stockrecords, attribute_values, and product_class, which I won't have at the time to update it. I could query ecommerce, then write to ecommerce, but it would be nice if there was a way to just update the title.

bill-filler commented on journals/apps/journals/handlers.py:
```diff
@@ -1,15 +1,26 @@
+page_published.connect(page_pub_receiver, sender=JournalPage)
+page_unpublished.connect(page_unpub_receiver, sender=JournalPage)
+page_published.connect(about_page_pub_receiver, sender=JournalAboutPage)
+page_unpublished.connect(about_page_unpub_receiver, sender=JournalAboutPage)
```
We also need to add listener for page deleted on the JournalAboutPage and update discovery to make status inactive.

tuchfarber replied:
We've discussed this offline and I believe the option was to leave this as is until we decide on a deletion/inactive strategy.

bill-filler replied:
@tuchfarber 
please add a story under the Beta epic to handle this case

bill-filler commented on journals/apps/journals/models.py:
```diff
@@ -320,6 +322,51 @@ def generate_basket_url(self, sku):
+        ecommerce_data = {
+            "title": self.title
+        }
+        update_service(
```
@tuchfarber 
Does this post fail because we only supply title? If so, remove this for now and please create a story for Beta2 to expose ecommerce API that allows changing only title.

tuchfarber commented:
New stories WL-1545 and WL-1546 added to backlog. Ecommerce patch removed.
